test/infrastructure/docker/internal/provisioning/cloudinit/writefiles.go

```diff
@@ -62,13 +66,81 @@ func newWriteFilesAction() action {
 	return &writeFilesAction{}
 }
 
-func (a *writeFilesAction) Unmarshal(userData []byte) error {
+func (a *writeFilesAction) Unmarshal(userData []byte, kindMapping kind.Mapping) error {
 	if err := yaml.Unmarshal(userData, a); err != nil {
 		return errors.Wrapf(err, "error parsing write_files action: %s", userData)
 	}
+	for i, f := range a.Files {
```

Comment:
It seems to me that adding all this code here, including `kindMapping kind.Mapping` adds a lot of unnecessary complexity to the solution, as this code has nothing to do with `writeFilesAction` functionality.

Given that we already have plenty of code duplicated between CloudInit and Ignition code for CAPD, I'd suggest to introduce some intermediate, common format which you can convert both CloudInit and Ignition custom data to to perform CAPD-specific mutations + deduplicate code responsible for converting it into provisioning commands.

Format can be as simple as needed, so something like:
```go
type File struct {
  Path string
  Mode string
  Contents string // Or []byte
}

type ProvisioningData struct {
  Files []File
  Commands []provisioning.Cmd
}
```

Then you can have a function `ProvisioningData -> ProvisioningData`, which will apply CAPD specific modifications and `[]File -> []provisioning.Cmd` function to convert all files into creation commands.

This also obliviously enables us to reuse flags modifications between cloud init and Ignition and makes it all very easy to test as it's all data-driven.

Reply:
Finding the best way forward for sharing common code if any is in scope of https://github.com/kubernetes-sigs/cluster-api/issues/8907

This PR scope is to get us in a place where we are less exposed to the impact of changes in the kindest/image, see https://github.com/kubernetes-sigs/cluster-api/issues/8788 for the full context